ksql-streams/src/test/java/io/confluent/ksql/execution/streams/StreamSelectKeyBuilderTest.java

```diff
@@ -0,0 +1,235 @@
+import io.confluent.ksql.execution.plan.ExecutionStepProperties;
+import io.confluent.ksql.execution.plan.StreamSelectKey;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.KeyValueMapper;
+import org.apache.kafka.streams.kstream.Predicate;
+import org.apache.kafka.streams.kstream.ValueMapperWithKey;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+
+public class StreamSelectKeyBuilderTest {
+  private static final String ALIAS = "ATL";
+  private static final LogicalSchema SCHEMA = LogicalSchema.builder()
+      .valueColumn("BIG", SqlTypes.BIGINT)
+      .valueColumn("BOI", SqlTypes.STRING)
+      .build()
+      .withAlias(ALIAS)
+      .withMetaAndKeyColsInValue();
+  private static final String KEY = "ATL.BOI";
+
+  @Mock
+  private KStream kstream;
+  @Mock
+  private KStream rekeyedKstream;
+  @Mock
+  private KStream filteredKStream;
+  @Mock
+  private KStream updatedKeyKStream;
+  @Mock
+  private ExecutionStep sourceStep;
+  @Captor
+  private ArgumentCaptor<Predicate> predicateCaptor;
+  @Captor
+  private ArgumentCaptor<KeyValueMapper> keyValueMapperCaptor;
+  @Captor
+  private ArgumentCaptor<ValueMapperWithKey> mapperCaptor;
+
+  private final QueryContext queryContext =
+      new QueryContext.Stacker(new QueryId("hey")).push("ya").getQueryContext();
+  private final ExecutionStepProperties properties = new DefaultExecutionStepProperties(
+      SCHEMA,
+      queryContext
+  );
+
+  private StreamSelectKey selectKey;
+
+  @Rule
+  public final MockitoRule mockitoRule = MockitoJUnit.rule();
+
+  @Before
+  @SuppressWarnings("unchecked")
+  public void init() {
+    when(sourceStep.getProperties()).thenReturn(properties);
+    when(kstream.filter(any())).thenReturn(filteredKStream);
+    when(filteredKStream.selectKey(any())).thenReturn(rekeyedKstream);
+    when(rekeyedKstream.mapValues(any(ValueMapperWithKey.class))).thenReturn(updatedKeyKStream);
+    givenUpdateRowkey();
+  }
+
+  @SuppressWarnings("unchecked")
+  private void givenUpdateRowkey() {
+    selectKey = new StreamSelectKey(
+        properties,
+        sourceStep,
+        KEY,
+        true
+    );
+  }
+
+  @SuppressWarnings("unchecked")
+  private void givenUpdateRowkeyFalse() {
+    selectKey = new StreamSelectKey(
+        properties,
+        sourceStep,
+        KEY,
+        false
+    );
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void shouldRekeyCorrectly() {
+    // When:
+    final KStream result = StreamSelectKeyBuilder.build(kstream, selectKey);
+
+    // Then:
+    verify(kstream).filter(any());
```

Comment:
These should be in order though right?   Consider using mockito's `InOrder` to test this.

```java
final InOrder inOrder = inOrder(kstream, filteredKStream, rekeyedKstream);
inOrder.verify(kstream).filter(any());
inOrder.verify(filteredKStream).selectKey(any());
inOrder.verify(rekeyedKstream).mapValues(any(ValueMapperWithKey.class));
```